### PR TITLE
added storage, but cannot get retrieve data

### PR DESCRIPTION
**Pull request by hannahsalameh**
I have added functionality to store comments. I cannot figure out why it is not printing, but I am asking on tech support right now.

### CODE REVIEW

**Comment by hannahsalameh** [comment]
wait i fixed it

**Comment by hannahsalameh** [comment]
http://hannahsalameh-step-2020.appspot.com/ link should be updated also!

File: walkthroughs/week-3-server/examples/server-stats/src/main/webapp/script2.js [hunk 1]
```diff
@@ -0,0 +1,41 @@
+// Copyright 2019 Google LLC
```
Comment (ofk-goog) [inline walkthroughs/week-3-server/examples/server-stats/src/main/webapp/script2.js:1]: This change seems unrelated to the comments functionality, so please move it to a different pull request

File: portfolio/src/main/webapp/script.js [hunk 2]
```diff
@@ -27,6 +27,33 @@ function getInfo(){
     })
 }
 
+function displayComment(){
+    console.log("function is running");
```
Comment (ofk-goog) [inline portfolio/src/main/webapp/script.js]: Please either remove this log line or say something more specific like "Displaying comments"

File: portfolio/src/main/webapp/designs.html [hunk 3]
```diff
@@ -13,6 +13,7 @@
         <div class="header_obj"><a href="projects.html" class="header_link"><h3> PROJECTS </h3></a></div>
         <div class="header_obj"><a href="designs.html" class="header_link"><h3> DESIGNS </h3></a></div>
         <div class="header_obj"><a href="about.html" class="header_link"><h3> ABOUT </h3></a></div>
+        <div class="header_obj"><a href="feedback.html" class="header_link"><h3> FEEDBACK </h3></a></div>
```
Comment (ofk-goog) [inline portfolio/src/main/webapp/designs.html:16]: It's becoming unruly to maintain a copy of this header_container in each HTML file. In a future PR you might want to centralize this if possible.

File: portfolio/src/main/java/com/google/sps/servlets/DataServlet.java [hunk 4]
```diff
@@ -14,32 +14,94 @@
...
 import java.io.IOException;
 import javax.servlet.annotation.WebServlet;
 import javax.servlet.http.HttpServlet;
 import com.google.gson.Gson;
 import java.util.ArrayList;
+import java.util.List;
 import javax.servlet.http.HttpServletRequest;
 import javax.servlet.http.HttpServletResponse;
 
 /** Servlet that returns some example content. TODO: modify this file to handle comments data */
```
Comment (ofk-goog) [inline portfolio/src/main/java/com/google/sps/servlets/DataServlet.java:32]: This comment is out of date

File: portfolio/src/main/java/com/google/sps/servlets/DataServlet.java [hunk 5]
```diff
@@ -14,32 +14,94 @@
...
-@WebServlet("/data")
+@WebServlet("/comments")
 public class DataServlet extends HttpServlet {
-
+    String json;
```
Comment (ofk-goog) [inline portfolio/src/main/java/com/google/sps/servlets/DataServlet.java]: Unused instance variable `json`, please delete

File: portfolio/src/main/java/com/google/sps/servlets/DataServlet.java [hunk 6]
```diff
@@ -14,32 +14,94 @@
...
+
+@Override
+public void doGet(HttpServletRequest request, HttpServletResponse response) throws IOException {
+    Query query = new Query("comment").addSort("timestamp", SortDirection.DESCENDING);
+    
+    DatastoreService datastore = DatastoreServiceFactory.getDatastoreService();
+    PreparedQuery results = datastore.prepare(query);
+
+    ArrayList<Comment> comments = new ArrayList<>();
+    for (Entity entity : results.asIterable()) {
```
Comment (ofk-goog) [inline portfolio/src/main/java/com/google/sps/servlets/DataServlet.java:91]: Consider writing a separate method to convert Entity to Comment

File: portfolio/src/main/webapp/feedback.html [hunk 7]
```diff
@@ -0,0 +1,31 @@
...
+        <div class="header_obj"><a href="projects.html" class="header_link"><h3> PROJECTS </h3></a></div>
+        <div class="header_obj"><a href="designs.html" class="header_link"><h3> DESIGNS </h3></a></div>
+        <div class="header_obj"><a href="about.html" class="header_link"><h3> ABOUT </h3></a></div>
+        <div class="header_obj"><a href="feedback.html" class="header_link"><h3> FEEDBACK </h3></a></div>
+  </header>
+  <body onload= "displayComment()">
+      <h1> Leave a Comment!</h1>
+      <p> Want Hannah to know what you think of her website? Feel free to leave a comment on this page to make sure she sees it! </p>
+      <form action= "/comments" method="POST">
+      <label for"name"> Name: </label><br>
```
Comment (ofk-goog) [inline portfolio/src/main/webapp/feedback.html]: missing `=`

File: portfolio/src/main/java/com/google/sps/servlets/DataServlet.java [hunk 8]
```diff
@@ -14,32 +14,94 @@
...
+        long timestamp = (long) entity.getProperty("timestamp");
+        long id = entity.getKey().getId();
+
+        Comment comment = new Comment(title, body, timestamp, id);
+        comments.add(comment);
+    }
+    Gson gson = new Gson();
+
+    response.setContentType("application/json;");
+    response.getWriter().println(gson.toJson(comments));
```
Comment (ofk-goog) [inline portfolio/src/main/java/com/google/sps/servlets/DataServlet.java:98]: The whole process of converting from datastore Entity to JSON here seems like magic. It's not obvious to me, as a reader, how the resulting JSON will look.

Normally I'd ask for a unit test to make sure the JSON looks how we expect. But the STEP program hasn't covered how to write those yet, have they?

Also, normally I'd say let's simplify this and produce the JSON without using Gson, since AFAICT Gson is designed for something we're not doing: turning Java objects into strings and then back into Java objects. But I think the STEP team asked you to use Gson?

**Review by ofk-goog — CHANGES_REQUESTED** [review]
This is a pretty big CL. It's not so much about number of lines of code, as it is about number of new App Engine technologies for Oliver to learn :)

File: walkthroughs/week-3-server/examples/server-stats/src/main/webapp/script2.js (hunk 1 above)
Comment (hannahsalameh) [reply walkthroughs/week-3-server/examples/server-stats/src/main/webapp/script2.js:1]: I'm not sure how to do that in this request, but I will avoid adding unrelated commits in the future

File: portfolio/src/main/java/com/google/sps/servlets/DataServlet.java (hunk 8 above)
Comment (hannahsalameh) [reply portfolio/src/main/java/com/google/sps/servlets/DataServlet.java:98]: No, the tutorial has not covered testing yet. I'm not sure if that's something I have not gotten to yet in the tutorial, or if that's going to be taught a different week. Should I look into creating unit tests?
Yeah the STEP tutorial has you using Gson for everything. I'm not sure why.

Comment (RegExrTech) [reply portfolio/src/main/java/com/google/sps/servlets/DataServlet.java:98]: Test driven development will be covered in week 5 of your internship (Project 4) so I wouldn't worry too much about it now, but feel free to start doing some preliminary research if you have time!

File: portfolio/src/main/webapp/designs.html (hunk 3 above)
Comment (hannahsalameh) [reply portfolio/src/main/webapp/designs.html:16]: I'll look into that for future PRs

**Review by ofk-goog — APPROVED** [review]
